### ISSUE TITLE
inference intervention和logits refine同时应用无法复现

### ISSUE THREAD

**Issue by Transl9** [LALBJ/PAI#4]
拜读了这篇具有启发性的文章后，尝试根据代码进行复现，在chairs数据集中遇到以下问题：

1. 文中 inference intervention 和 logits refine 两个模块按我理解是同时生效的，但是在`attention.py`的line 90写道`if use_attn and not use_cfg`，这个判断条件是否意味着使用 logits refine 后， inference intervention 就不再生效？
2. 尝试将上述代码中的`if use_attn and not use_cfg`改为`if use_attn`后，可以同时使用 inference intervention 和 logits refine 两个模块了，但是按照文章提供的细节设置，仅使用 inference intervention 的结果如下：

```

> chair_eval_layers_2-32_tokens_512_bs_1_beams_5_attn_0.5.json

  "overall_metrics": {
    "CHAIRs": 0.248,
    "CHAIRi": 0.09473292387865252,
    "Recall": 0.6652119700748129,
    "Precision": 0.8660714285714286,
    "F1": 0.7524682651622002,
    "Len": 1.2216600000000002
  }
```

同时使用 inference intervention 和 logits refine 的结果反而下降了，达不到文中汇报的性能：

```

> chair_eval_layers_2-32_tokens_512_bs_1_beams_5_attn_0.5_cfg_1.1.json

  "overall_metrics": {
    "CHAIRs": 0.276,
    "CHAIRi": 0.0881516587677725,
    "Recall": 0.649002493765586,
    "Precision": 0.8497959183673469,
    "F1": 0.7359490986214211,
    "Len": 1.00884
  }
```

3. 请问如何设置 Greedy 和 Nucleus 解码方式？

再次感谢作者对社区作出的贡献！

**Comment by LALBJ** [comment]
感谢对于我们工作的关注和认可！我依次回答一下你提的问题：

1. 首先，inference intervention 和 logits refine 是两个独立的组件，你可以任意去单独使用或者一起使用。但是在具体实现中，实现 logits refine 组件时，需要用带有图像 input 输出的 logits 去减去不带图像 input 输出的 logits，在计算不带图像 input 输出的 logits 时，是不需要调用 inference intervention 组件的，所以采用 `if use_attn and not use_cfg` 这一逻辑判断。
2. 按照在 1 中的解释，所以你这里将代码修改是有问题的。除此之外，CHAIR 评估我们采用了与 OPERA 相同的 setting，即随机采样了 500 个 COCO 数据样本进行评估，这种 setting 也会造成结果的浮动。所以，我们后续会公布我们评估所用的样本。
3. 将 beam 设置为 1 就会调用 Greedy 解码，设置 args.sample 为 True 则会调用 Nucleus Sample 解码。

**Comment by Transl9** [comment]
感谢你们及时的回复与解释！我之前的问题得到了解决。

我想再请教一下，论文的Fig.5可视化了 <BOS> token的注意力模式，请问<BOS> token的具体位置在token序列的什么地方？按我的理解，自回归任务中参与forward的token种类依次是system, visual, instrucition, history output，请问<BOS> token的位置是在history output的第一个token吗？
另外，请问能够分享Fig.5可视化的代码吗？

再次感谢！

**Comment by LALBJ** [comment]
哈喽，

绘制 Fig.5 时，我们对 inference 过程中，当前生成 token 分配在 <BOS> token 上的 attention weights 比例进行了收集。然后，对多个样本中的多个 token 取平均得到了 Fig.5 的结果。也就是对收集到的 (sample_size, sequence_length, layers, heads, 1) 的 tensor 取平均得到 (layers, heads, 1) 后绘制，即为 Fig.5 的结果。

后续我们也会开源可视化代码，如果有别的问题欢迎提问讨论。

**Comment by Transl9** [comment]
感谢回复！

另外请问如果要添加对其他开源VLM的应用，需要在哪些代码中进行修改？例如OPERA论文中使用的InstructBLIP模型

**Comment by LALBJ** [comment]
抱歉，回复的稍微有些晚了。

如果想要在别的模型上支持我们的方法，你可以直接参考 `model_loader.py` 在我们的 codebase 上支持新的模型。也可以导出我们的算法组件在别的 codebase 上复现，对于 inference intervention 组件建议参考 `attention.py` 中的 `llama_modify` 函数。对于 logits refine 组件，建议参考 `model_loader.py` 中 `init_cfg_processor` 函数。

关于，可视化代码和 CHAIR 评估子集由于我已从原实验室毕业，所以数据被清除了，我们会尽快复现并开源这些内容。

**Comment by pandawynn** [comment]
> 抱歉，回复的稍微有些晚了。
> 
> 如果想要在别的模型上支持我们的方法，你可以直接参考 `model_loader.py` 在我们的 codebase 上支持新的模型。也可以导出我们的算法组件在别的 codebase 上复现，对于 inference intervention 组件建议参考 `attention.py` 中的 `llama_modify` 函数。对于 logits refine 组件，建议参考 `model_loader.py` 中 `init_cfg_processor` 函数。
> 
> 关于，可视化代码和 CHAIR 评估子集由于我已从原实验室毕业，所以数据被清除了，我们会尽快复现并开源这些内容。

Is the code for visualization available now？

**Comment by sreetamasarkar** [comment]
Sorry, what should be the input arguments if we want to use inference intervention and logits_refine together? Am I correct in understanding that if args.use_cfg is True, then inference intervention is turned off?
